flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesResourceManagerTest.java

```diff
@@ -0,0 +1,366 @@
+			return callAsync(callable, TIMEOUT);
+		}
+
+		@Override
+		protected void runAsync(Runnable runnable) {
+			runnable.run();
+		}
+
+		MainThreadExecutor getMainThreadExecutorForTesting() {
+			return super.getMainThreadExecutor();
+		}
+
+		SlotManager getSlotManager() {
+			return this.slotManager;
+		}
+	}
+
+	@Test
+	public void testStartAndStopWorker() throws Exception {
+
+		final TestingKubernetesResourceManager resourceManager = this.createAndStartResourceManager(flinkConfig);
+		registerSlotRequest(resourceManager);
+
+		final KubernetesClient client = getKubeClient();
+		final PodList list = client.pods().list();
+		assertEquals(1, list.getItems().size());
+		final Pod pod = list.getItems().get(0);
+
+		final Map<String, String> labels = getCommonLabels();
+		labels.put(Constants.LABEL_COMPONENT_KEY, Constants.LABEL_COMPONENT_TASK_MANAGER);
+		assertEquals(labels, pod.getMetadata().getLabels());
+
+		assertEquals(1, pod.getSpec().getContainers().size());
+		final Container tmContainer = pod.getSpec().getContainers().get(0);
+		assertEquals(CONTAINER_IMAGE, tmContainer.getImage());
+
+		final String podName = CLUSTER_ID + "-taskmanager-1-1";
+		assertEquals(podName, pod.getMetadata().getName());
+
+		// Check environments
+		assertThat(tmContainer.getEnv(), Matchers.contains(
+			new EnvVarBuilder().withName(Constants.ENV_FLINK_POD_NAME).withValue(podName).build()));
+
+		// Check task manager main class args.
+		assertEquals(3, tmContainer.getArgs().size());
+		final String confDirOption = "--configDir " + flinkConfig.getString(KubernetesConfigOptions.FLINK_CONF_DIR);
+		assertTrue(tmContainer.getArgs().get(2).contains(confDirOption));
+
+		resourceManager.onAdded(Collections.singletonList(new KubernetesPod(flinkConfig, pod)));
+		final ResourceID resourceID = new ResourceID(podName);
+		assertThat(resourceManager.getWorkerNodeMap().keySet(), Matchers.contains(resourceID));
+
+		registerTaskExecutor(resourceManager, resourceID);
+
+		// Unregister all task executors and release all task managers.
+		CompletableFuture<?> unregisterAndReleaseFuture = resourceManager.runInMainThread(() -> {
+			resourceManager.getSlotManager().unregisterTaskManagersAndReleaseResources();
+			return null;
+		});
+		unregisterAndReleaseFuture.get();
+		assertEquals(0, client.pods().list().getItems().size());
+		assertEquals(0, resourceManager.getWorkerNodeMap().size());
+
+		resourceManager.getRpcService().stopService().get();
+	}
+
+	@Test
+	public void testTaskManagerPodTerminated() throws Exception {
+		final TestingKubernetesResourceManager resourceManager = this.createAndStartResourceManager(flinkConfig);
+		registerSlotRequest(resourceManager);
+
+		final KubernetesClient client = getKubeClient();
+		assertEquals(1, client.pods().list().getItems().size());
+		final Pod pod = client.pods().list().getItems().get(0);
+		final String taskManagerPrefix = CLUSTER_ID + "-taskmanager-1-";
+		assertEquals(taskManagerPrefix + 1, pod.getMetadata().getName());
```

Comment:
This was already verified in above test, so no need to double verify again, also for `assertEquals(1, client.pods().list().getItems().size())`

Reply:
I will remove the unnecessary verification. For `assertEquals(1, client.pods().list().getItems().size())`, we should not remove all. Because if we `onModified` or `Prepare pod of previous attempt`, we need to check the pod number.

Reply:
exactly